File: appinventor/components/src/com/google/appinventor/components/common/OptionList.java
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright 2020 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package com.google.appinventor.components.common;

/**
 * A marker interface for defining option list helper blocks via the Java scripting language.
 * This can be expanded later if we want to get more information out of option list definitions.
 */
public interface OptionList<T> {
  /**
   * Returns the underlying value of the option.
   */
  T toUnderlyingValue();
}

File: appinventor/components-ios/src/OptionList.swift
// -*- mode: swift; swift-mode:basic-offset: 2; -*-
// Copyright 2023 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

import Foundation

@objc public protocol OptionList {
  func toUnderlyingValue() -> AnyObject
}

/**
 * Generates a lookup table for classes that implement the `OptionList` protocol.
 *
 * - Parameter options: A varargs list of options
 * - Returns: A dictionary mapping the underlying value of each option to the option itself.
 */
func generateOptionsLookup<Key, T: OptionList>(_ options: T...) -> [Key:T] {
  var result = [Key:T]()
  options.forEach { item in
    guard let key = item.toUnderlyingValue() as? Key else {
      return
    }
    result[key] = item
  }
  return result
}
